src/main/scala/com/daml/projection/Migration.scala

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.projection
+
+import akka.actor.ActorSystem
+import com.typesafe.scalalogging.StrictLogging
+import org.flywaydb.core.Flyway
+
+import scala.jdk.CollectionConverters._
+import scala.util.Try
+import java.io.PrintWriter
+import java.{ sql, util }
+import java.sql.{
+  Blob,
+  CallableStatement,
+  Clob,
+  Connection,
+  DatabaseMetaData,
+  NClob,
+  PreparedStatement,
+  SQLWarning,
+  SQLXML,
+  Savepoint,
+  Statement,
+  Struct
+}
+import java.util.Properties
+import java.util.concurrent.Executor
+import java.util.logging.Logger
+
+private object Migration extends StrictLogging {
+  def projectionTableName(implicit sys: ActorSystem) =
+    sys.settings.config.getString("projection.projection-table-name")
+
+  def migrateIfConfigured(con: java.sql.Connection)(implicit sys: ActorSystem): Int = {
+    val migrateOnStart = sys.settings.config.getBoolean("projection.flyway.migrate-on-start")
+    val internalLocations = sys.settings.config.getStringList("projection.flyway.internal-locations").asScala.toList
+    val userProvidedLocations = sys.settings.config.getStringList("projection.flyway.locations").asScala.toList
+    val flywayLocations = internalLocations ++ userProvidedLocations
+    val flyway = Flyway.configure()
+      .placeholders(
+        Map(
+          "projection_table_name" -> projectionTableName
+        ).asJava
+      )
+      .dataSource(new SingleConnectionDataSource(con))
```

Comment:
If you derive your connection from `com.daml.dbutils.ConnectionPool.connect`, this supplies a `DataSource` without holes in it. Perhaps this `DataSource` could be the root source of connections as well?

Reply:
Updated so users can use any `DataSource`.